### PR TITLE
AC-480 Patient model should extend Person

### PR DESCRIPTION
**Pull request by juniorjainsahab**
## Description of what I changed
* Changing the relationship between `Patient` and `Person` to `is a` relationship instead of `has a` relationship.
* Details that can be updated in a patient now can be updated through a new `PatientDetails` object.
* We had a hard dependency on the backend API while updating or registering patient, It still accepts certain detail as part of `person` key. So we have exposed a `getPerson` method while making any network request.  

## Issue I worked on
After this PR, `Patient` model is extending Person. From now onwards, we have to get patient details from the Patient itself instead of from Person.
## Before
     patient.getPerson().getNames();
## After
     patient.getNames();

I have already run the unit test at my local machine and all are passing successfully and I also run the application on the device and it also works the expected way.

JIRA Issue: https://issues.openmrs.org/browse/AC-480

## Checklist: I completed these to help reviewers :)

- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [X] All new and existing **tests passed**.

- [X] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientPresenter.java [hunk 1]
```diff
@@ -20,8 +20,11 @@
 import org.openmrs.mobile.api.retrofit.PatientApi;
 import org.openmrs.mobile.dao.PatientDAO;
 import org.openmrs.mobile.listeners.retrofit.DefaultResponseCallbackListener;
+import org.openmrs.mobile.models.IdentifierType;
 import org.openmrs.mobile.models.Module;
 import org.openmrs.mobile.models.Patient;
+import org.openmrs.mobile.models.PatientDto;
+import org.openmrs.mobile.models.Person;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientPresenter.java]: @juniorjainsahab Please remove all unused imports.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/Patient.java [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import com.google.gson.annotations.Expose;
 import com.google.gson.annotations.SerializedName;
 
+import org.openmrs.mobile.dao.PatientDAO;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/Patient.java]: We shouldn't need a DAO import in a model class.

**Comment by juniorjainsahab** [comment]
Creating setters for multiple fields in a class can be dangerous sometimes, One may forget to set some fields which will result in the `run time` failure. These things can be easily resolved at the language level on the compile time by setting up the `data contract` which a class needs to be initialized with. `PatientDetails` is one such class.

Regarding `PatientDto`, this class is something which is defining the structure with backend APIs, If i remove it and use `Patient` instead. backend server will send `400 bad request` since the structure of request body has changed. 

Let me know if you think otherwise and have a suggestion on those two points. Meanwhile, I'll be fixing the unused imports issue.

**Comment by f4ww4z** [comment]
@juniorjainsahab What you can do is make a constructor in `Patient`, as currently the way to create a Patient object is create an empty `Patient` object and calling multiple setters. See [an example ](https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/Patient.java#L54)in the openmrs-core repo. You can edit the constructor to take in the same variables as in the `PatientDetails` class, so creating a `Patient` object with arguments defined in the constructor is the same as creating an empty object and calling setters on it. You can also overload the constructor, if necessary.

I think that `PatientDto` looks good to me, just add a javadoc in the class and describe how the class is used.

**Comment by juniorjainsahab** [comment]
@f4ww4z all the changes are done.

**Comment by juniorjainsahab** [comment]
@f4ww4z 
if you have some feedback do let me know otherwise please merge it as soon as possible because it will be hard to remove merge conflicts.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+
+    /**
+     *
+     * @return
+     *     The patient
+     */
+    public Patient getPatient(){
+        Patient patient = new Patient();
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: @juniorjainsahab I requested to make a `Patient` constructor, so you have not done that part. The constructor is very important so that in the future we do not forget to set some attributes. It also makes the code more concise.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/Person.java [hunk 4]
```diff
@@ -45,9 +45,7 @@
     private Bitmap photo;
 
     /**
-     * 
-     * @return
-     *     The names
+     * @return The names
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/Person.java]: This doesn't improve the documentation. Revert the changes to this **whole file** or properly describe the variables.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
+package org.openmrs.mobile.models;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: Add the OpenMRS license header at the start.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PatientDto {
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: I mentioned describing this class with a javadoc.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    @SerializedName("person")
+    @Expose
+    private Person person;
+
+    /**
+     *
+     * @return
+     *     The uuid
+     */
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: This doesn't improve the documentation. You can either describe the variable or remove the javadoc. Do this for all the method docs in this file.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java (hunk 3 above)
Comment (juniorjainsahab) [reply openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: this looks like a really big change since `AddEditPatientFragment` holds one of the patient object. and the very same object keeps updating every time when the user clicks a submit button.

For an issue like this where we just had to extend Patient with Person and establish an `is a` relationship. these kinds of change may totally change the app's behavior and can cause the app to crash and that's why application again needs to be thoroughly tested. We have already change the create/update API's payload to a different object `PatientDto` and that works in an expected manner.

I'd recommend to not to create constructor this time and fix it in an iterative manner. Could you please create a new issue for that, I'd be happy to pick that up and fix all mutability issue.

Let me know if you think otherwise.

**Comment by juniorjainsahab** [comment]
@f4ww4z , I think there has been some change regarding play store stuff and causing both of the build services to fail. As far as I know, this is not related to my changes. Could you please verify that?

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java (hunk 3 above)
Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: I now realize that this change will be bigger than expected, but still an issue that needs fixing. @juniorjainsahab do file a new issue for this, since I may forget.

**Comment by juniorjainsahab** [comment]
@f4ww4z I have squashed all the commits into one. Please let me know if something is required from my side.

File: openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java (hunk 3 above)
Comment (juniorjainsahab) [reply openmrs-client/src/main/java/org/openmrs/mobile/models/PatientDto.java]: @f4ww4z new issue is also created you can check it on [AC-617](https://issues.openmrs.org/browse/AC-617).

**Comment by f4ww4z** [comment]
@juniorjainsahab The external CI error is now fixed - please pull the latest changes from master.

**Comment by juniorjainsahab** [comment]
@f4ww4z now both the ci builds are passing. So it is now ready for merge.
